### ISSUE TITLE
Support matching of failure on steps following an `around` step

### ISSUE THREAD

**Issue by timriley** [dry-rb/dry-transaction#99]
See @grabr's discussion of the issue here: https://discourse.dry-rb.org/t/dry-transaction-pattern-matching-of-steps-inside-the-around-step/514/1

@flash-gordon do you have any initial thoughts on this?

**Comment by timriley** [comment]
My preferred approach here would be to keep the existing pattern match API. So for a transaction like this:

```ruby
class MyTrans
  step :outer
  step :around
  step :inner
end
```

I'd like to be able to do this to match on a failure in `:inner`:

```ruby
my_trans.(input) do |r|
  r.failure(:inner) do |f|
    # ...
  end
end
```

**Comment by flash-gordon** [comment]
🤔🤔🤔 this looks like a bug to me, technically, `around` steps can never fail. Probably, the bug caused by an _unexpected_ outcome of composition where initial information gets overridden, but I didn't try to verify my guess

**Comment by Hugo-Hache** [comment]
I faced the same problem trying to catch a specific step in a call block, but was unable to do this because of an around step. I digged into the codebase, and it looks to me the bug originates from https://github.com/dry-rb/dry-transaction/blob/406a188e127a08fb0a3e6b944587448d5b25c379/lib/dry/transaction/result_matcher.rb#L13

By replacing `result.left.step.step_name == step_name` by `result.left.value.step.step_name == step_name` the matching would happen on the step name rather the around. 

@flash-gordon let me know if it looks good for you, if so I'll submit a PR.

**Comment by timriley** [comment]
@Hugo-Hache That'd certainly fix the issue for transactions with `around` steps, but I don't think it'd work for everything. I think we might need to built a special StepFailure object specifically for `around` steps that exposes the name of the failing step within it. To make this tidy, we might need to come up with a system that allows each step adapter to provide its own StepFailure object (with most of them using the default, except `around`).

Would you be up for giving something like that a go? (@flash-gordon — that seem reasonable to you?)

**Comment by GustavoCaso** [comment]
Fixed in #106
